Make primitive tests go back and forth, fix float params in imported functions

kajacx opened the pull request:
Adding tests that go back and forth makes the testing more robust, as it tests the interaction of a larger part of the system.

Addresses #180, but there is still more to do, for example generating a warning when an imported function takes `f32`/`f64`, or marking it as deprecated with a link to this issue.

kajacx commented:
I'm not sure if I can "pause" the PR, I've had an idea to encode the floats as integers to avoid the issue, since there is a conversion already going on with the `WasmAbi::AbiType` type.

But it's only being converted on the host and not in the plugin. Not sure if that is intentional or why is that.

arendjr commented on examples/example-protocol/src/main.rs:
```diff
@@ -48,17 +48,20 @@ fp_import! {
+    // FIXME: Imported functions get 0.0 instead of the float values when called from a plugin.
+    // This is a bug in wasmer 2.3, but can be worked around by using a one-sized array instead.
+    // See https://github.com/fiberplane/fp-bindgen/issues/180
+    fn import_primitive_f32_add_one(arg: [f32; 1]) -> f32;
+    fn import_primitive_f64_add_one(arg: [f64; 1]) -> f64;
```
I think it may make sense to keep the non-array versions too, but simply omit those from the Wasmer test with a comment similar to this one explaining why. Then when we add a Wasmer 3 runtime, it can have reinstate the tests for those.

kajacx replied:
Yea, that's what I was thinking too.

kajacx replied:
But with the "pass floats as integer" angle, you can pass floats around directly (at least as far as the functions *you* will be working with are concerned) without the need for the "one-sized array" hack, which I think could be better.

arendjr reviewed (approved):
You'll just need to update some `expected_*` files to get the failing tests to pass.

Not entirely what you mean by encoding floats as integers? Do you intend to transmute the types?

kajacx commented:
> Not entirely what you mean by encoding floats as integers? Do you intend to transmute the types?

Not directly transmute, but yes. I tried this:

```rust
impl WasmAbi for f32 {
    type AbiType = u32;

    #[inline]
    fn to_abi(self) -> Self::AbiType {
        u32::from_le_bytes(self.to_le_bytes())
    }

    #[inline]
    fn from_abi(value: Self::AbiType) -> Self {
        Self::from_le_bytes(value.to_le_bytes())
    }
}
```

but it didn't work. Wasmer fails at runtime, saying that the function signatures don't match:

```
---- test::primitives stdout ----
thread 'test::primitives' panicked at 'called `Result::unwrap()` on an `Err` value:
Link(Import("fp", "__fp_gen_import_primitive_f32_add_one",
IncompatibleType(Function(FunctionType { params: [I64], results: [F32] }),
Function(FunctionType { params: [I64], results: [I32] }))))',
src\spec\bindings.rs:29:63
```

After some digging, I found out that the `WasmAbi` type conversion is done only on host, but not in the plugin. I tried to draw it into a picture:
![Exanging types in wasm and converting them to their ab itype](https://user-images.githubusercontent.com/5281491/224763839-b4110b52-0e9a-4c46-8add-e8520297eaa9.png)

I don't know why the plugin "glue" code doesn't convert the primitive types to their abi types like the host code does. Maybe it is an oversight? What confuses me is that it works for `bool`, whose abi type is `u8`. I tried to change it to `u16` and that works too. I guess wasmer converts it for you? I will have to try that.

kajacx commented:
Ok, more research showed me that the abi conversion is not needed on the plugin side, because `bool` is converted to `u8` automatically. Somehow. Although it is done so by "modulo 2" rather than equality with `0`, but I guess that doesn't matter.

So I see 2 options now:

1) Add the abi conversions on the plugin side as well. This would allow to pass floats and integers and bypass the wasmer float argument issue
2) Keep the code as it is, including the "array" workaround as an example of passing floats into an imported function

kajacx commented:
Ok, I have updated the PR. I have split the float methods into 2: one that uses the workaround, and one that doesn't:
```rust
fn import_primitive_f32_add_one(arg: f32) -> f32;
fn import_primitive_f64_add_one(arg: f64) -> f64;

fn import_primitive_f32_add_one_wasmer2(arg: [f32; 1]) -> f32;
fn import_primitive_f64_add_one_wasmer2(arg: [f64; 1]) -> f64;

fn export_primitive_f32_add_three(arg: f32) -> f32;
fn export_primitive_f64_add_three(arg: f64) -> f64;

fn export_primitive_f32_add_three_wasmer2(arg: f32) -> f32;
fn export_primitive_f64_add_three_wasmer2(arg: f64) -> f64;
```
but because all both the wasmer and deno runtimes are using the same protocol and plugin, i had to implement the "wasmer2 workaround" imported methods in the ts runtime as well. So I added tests for them too, because why not.

Also, I have updated the generated files, so it should look good now.

arendjr commented:
Please add a CHANGELOG entry (feel free to mention yourself ;)) and I'll merge it!

arendjr commented:
:tada:
